src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentPrerenderer.cs

```diff
@@ -101,29 +101,30 @@ internal sealed class ComponentPrerenderer
         }
     }
 
-    private static ServerComponentInvocationSequence GetOrCreateInvocationId(ViewContext viewContext)
+    private static ServerComponentInvocationSequence GetOrCreateInvocationId(ActionContext actionContext)
     {
-        if (!viewContext.Items.TryGetValue(ComponentSequenceKey, out var result))
+        if (!actionContext.HttpContext.Items.TryGetValue(ComponentSequenceKey, out var result))
         {
             result = new ServerComponentInvocationSequence();
-            viewContext.Items[ComponentSequenceKey] = result;
+            actionContext.HttpContext.Items[ComponentSequenceKey] = result;
         }
 
         return (ServerComponentInvocationSequence)result;
     }
 
     // Internal for test only
-    internal static void UpdateSaveStateRenderMode(ViewContext viewContext, RenderMode mode)
+    internal static void UpdateSaveStateRenderMode(ActionContext actionContext, RenderMode mode)
     {
+        // TODO: This will all have to change when we support multiple render modes in the same response
         if (mode == RenderMode.ServerPrerendered || mode == RenderMode.WebAssemblyPrerendered)
         {
-            if (!viewContext.Items.TryGetValue(InvokedRenderModesKey, out var result))
+            if (!actionContext.HttpContext.Items.TryGetValue(InvokedRenderModesKey, out var result))
```

Comment:
I don't remember the exact reason for this, but I believe that we were putting it inside the viewcontext for a specific reason. Might be good to chase the details.

The main difference is that the viewcontext changes when rendering partials, etc. (kind of like a cascading value). While the action context doesn't.

Reply:
I could create a ViewContext to use here but it would be totally fake - the ViewData etc would not contain anything. I don't think we want to encourage use of ViewContext concepts inside Razor components as that makes them unable also to be used interactively.

In this case, `InvokedRenderModesKey` is a way of auto-detecting which render mode you're using across all the component tag helpers in your page. But:

 * Today, we require you to use just one consistently across the whole response. It's not possible to change it on a per-partial basis.
 * By the time 8.0 ships, we have to eliminate this restriction completely anyway because we want to support inclusion of any combination of component types

So, I don't think we need to track this info on a per-partial basis. If there was another reason I'm not aware of it but please let me know if you can find any memories of it!